blockscout-ens/bens-logic/src/subgraph/ens/wildcard.rs

```diff
@@ -0,0 +1,186 @@
+                return Ok(None);
+            } else {
+                // domain is resolved with wildcard, so we need to recheck it
+                tracing::info!(
+                    domain_id = domain.id,
+                    domain_name = domain.name,
+                    "found domain with wildcard-resolution, rechecking it"
+                )
+            }
+        }
+    };
+
+    let result = ccip_read::call_to_resolver(from_user, resolver_address)
+        .await
+        .context("resolve using ccip call")?;
+
+    if !result.addr.is_zero() {
+        Ok(Some(creation_domain_from_rpc_resolution(
+            from_user,
+            result,
+            maybe_existing_domain,
+        )))
+    } else {
+        Ok(None)
+    }
+}
+
+async fn any_resolver(
+    db: &PgPool,
+    from_user: &DomainNameOnProtocol<'_>,
+) -> Result<Option<(Address, Option<Domain>)>, anyhow::Error> {
+    let protocol = from_user.deployed_protocol.protocol;
+    let name_options = from_user
+        .inner
+        .iter_parents_with_self()
+        .map(|name| DomainNameOnProtocol::new(name, from_user.deployed_protocol))
+        .collect::<Vec<_>>();
+    // try to find resolver in db
+    let maybe_domain_with_resolver = any_resolver_in_db(db, name_options.clone()).await?;
+    if let Some((resolver, found_domain)) = maybe_domain_with_resolver {
+        Ok(Some((resolver.resolver_address, Some(found_domain))))
+    } else if protocol.info.registry_contract.is_some() || protocol.info.network_id == 1 {
+        // try to find resolver on chain.
+        // if custom registry is set, we can try to find resolver in registry
+        if let Some(resolver_address) = any_resolver_rpc(name_options).await {
+            Ok(Some((resolver_address, None)))
+        } else {
+            Ok(None)
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+async fn any_resolver_in_db(
+    db: &PgPool,
+    names: Vec<DomainNameOnProtocol<'_>>,
+) -> Result<Option<(ResolverInSubgraph, Domain)>, anyhow::Error> {
+    let found_parent_domains = {
+        let input = subgraph::sql::FindDomainsInput::Names(names.clone());
+        let only_active = false;
+        let pagination = None;
+        subgraph::sql::find_domains(db, input, only_active, pagination)
+            .await
+            .context("searching parents in db")?
+    };
+
+    let result = names
+        .iter()
+        .filter_map(|name| {
+            found_parent_domains
+                .iter()
+                .find(|d| d.name.as_ref() == Some(&name.inner.name))
+                .and_then(|domain| {
+                    let maybe_resolver = domain
+                        .resolver
+                        .as_ref()
+                        .and_then(|r| subgraph::ResolverInSubgraph::from_str(r).ok());
+                    maybe_resolver.map(|r| (r, domain.clone()))
+                })
+        })
+        .next();
+    Ok(result)
```

Comment:
Nit: using explicit for cycle may be more readable in that case (imo, it better conveys the intention to return the first non-empty result)

```suggestion
    for name in names {
        let maybe_domain = found_parent_domains
            .iter()
            .find(|d| d.name.as_ref() == Some(&name.inner.name));

        if let Some(domain) = maybe_domain {
            let maybe_resolver = domain
                .resolver
                .as_ref()
                .and_then(|r| ResolverInSubgraph::from_str(r).ok());

            if let Some(resolver) = maybe_resolver {
                return Ok(Some((resolver, domain.clone())));
            }
        }
    }

    Ok(None)
```